### ISSUE TITLE
Cannot dismiss a wallet user interface and open a different one

### ISSUE THREAD

**Issue by lorenzopolidori** [reach-sh/reach-lang#755]
**Describe the error**
A web application shows a user interface element to let users select which wallet fallback to use. 

* If users initially select MyAlgoWallet, a pop-up window is displayed to enter the password. Users can dismiss the pop-up and select PeraWallet instead. If they do so, the WalletConnect QR-code interface is not displayed, but instead MyAlgoWallet pop-up window is displayed again.

* If users initially select PeraWallet, the WalletConnect QR-code interface is displayed. Users can dismiss the QR-code interface and select MyAlgoWallet instead. If they do so, the MyAlgoWallet pop-up window is not displayed.

**To Reproduce**

Each UI button triggers either one of these two:
```js
async function connectMyAlgoWallet() {
  reach.setWalletFallback(reach.walletFallback({
    providerEnv: 'TestNet', MyAlgoConnect
  }))
  await connectWallet()
}
```
or
```js
async function connectPeraWallet() {
  reach.setWalletFallback(reach.walletFallback({
    providerEnv: 'TestNet', WalletConnect
  }))
  await connectWallet()
}
```
where `connectWallet()` is defined as follows:
```js
async function connectWallet() {
    let account;
    try {
      account = await reach.getDefaultAccount()
    } catch (e) {
      console.error('Error when connecting to wallet, or connection cancelled', e)
      return;
    }
    const balance = await reach.balanceOf(account)
    reach.formatCurrency(balance, 4)
}
```
* If `connectMyAlgoWallet()` is called first, the program stops at `await reach.getDefaultAccount()`, and the MyAlgoWallet pop-up window is displayed. If users dismiss the pop-up window, `reach.getDefaultAccount()` throws an exception, which is caught to report the event to the user. Users can then connect to PeraWallet from the UI selector, and `connectPeraWallet()` is called. However, `reach.getDefaultAccount()` does not display the WalletConnect QR-code interface, but instead the MyAlgoWallet pop-up window is displayed again.
* If `connectPeraWallet()` is called first, the program stops at `await reach.getDefaultAccount()`, and the WalletConnect QR-code interface is displayed. If users dismiss the QR-code interface, `await reach.getDefaultAccount()` does not return, nor it throws an exception (but it probably should, as it happens when cancelling MyAlgoWallet). The program is essentially stuck. Users can then choose MyAlgoWallet from the UI selector, and `connectMyAlgoWallet()` is called. However, nothing happens in this case.

**Expected behavior**

Users should be able to dismiss a Wallet interface and open a different one successfully.

**Extra information**

Reach `0.1.9`
Reach stdlib `0.1.9-rc.6`

**Comment by jeapostrophe** [comment]
`setWalletFallback` does not change anything about the wallet ever. The Reach standard library checks one time whether your browser has a compliant ARC11 wallet. If it doesn't, then it runs the wallet fallback, once, to create one. I think you want to call `loadStdlib` a second time and get a different standard library instance. Perhaps @MrJackdaw can give some insight into what they did in Humble.

**Comment by lorenzopolidori** [comment]
Do you think it would not be possible to change this behaviour and run the wallet fallback when calling `getDefaultAccount()` every time after `setWalletFallback()` is called?

Also, even keeping the current behaviour, a bug still stands, as I reported above. You cannot open WalletConnect, dismiss it with the X on its interface, and then call it again. When the WalletConnect interface is dismissed, there is no way to know that happened in the code, as `getDefaultAccount()` does not throw an exception.

**Comment by jeapostrophe** [comment]
The wallet is used for more than just `getDefaultAccount`. An ARC11 wallet is responsible for controlling access to the network as well, so you use it for making any query. It really is for all interactions with `stdlib`.

As far as the WalletConnect problem, I believe this is a problem with WalletConnect itself. I don't believe the code we're calling does throw an exception or tell us in anyway that the user clicked X. The code is pretty simple that we're using to interact with the WC library --- https://github.com/reach-sh/reach-lang/blob/master/js/stdlib/ts/ALGO_WalletConnect.ts

The documentation for WC is here --- https://docs.walletconnect.com/client-api

You might think that `disconnect` means they clicked X, but I think it means that you told it to disconnect. Similarly, there's `session_request`, but I think that's when WE ask it to make the session. I don't know if there's a way to know about canceling the WC request.

**Comment by lorenzopolidori** [comment]
@jeapostrophe FYI, I have tried to call `loadStdlib` every time before I want to connect to a wallet (i.e. call `getDefaultAccount`), but the issue is still there - I can't cancel a wallet selection nor cancel WalletConnect and open it again. 

This is my simple test code for reference: https://github.com/lorenzopolidori/nextjs-reach-wallet-connect/blob/master/pages/index.js

Happy to help on this issue, I think we need to find a viable solution to implement a wallet picker. 
Tinyman, for example, seems to work as expected: https://testnet.tinyman.org

**Comment by jeapostrophe** [comment]
@MrJackdaw Can you comment here, please?

**Comment by grzracz** [comment]
Would also love to get this solved, this also blocks any chance of disconnecting wallets. The only way to do it currently is to refresh the page which feels very unprofessional

**Comment by lorenzopolidori** [comment]
Any development on this issue?

**Comment by MrJackdaw** [comment]
Sorry about the belated response, all. Just caught up on the thread. 

First though,
#### The (anti-climactic) short answer
* Reload the window before attempting to reconfigure `stdlib` settings (e.g. current blockchain or network provider). 

This is how we handle it on HumbleSwap. 

#### The slightly longer explanation 
As Jay noted, setting a `walletFallback` effectively tells `stdlib` 

> This thing I just gave you will be (e.g.) `window.algorand` from now on.

So it becomes the fallback until you clear it.
And right now, the only way to clear that (and any) data from an original configuration is to reload the window. 

TLDR: **Setting a walletFallback happens once per window session** (i.e. until you refresh the page).

**Comment by lorenzopolidori** [comment]
@MrJackdaw, thanks. 

Out of curiosity, are you telling the user to reload the page manually, or are you doing it automatically with `location.reload()`?

In general, would it not make sense to implement a more flexible approach in `stdlib` where we can reset the wallet fallback if needed?

**Comment by jeapostrophe** [comment]
It is now possible to load independent instances of the standard library easily...

https://docs.reach.sh/frontend/#ref-frontends-js-loader

And we've added a discussion about how to use `setWalletFallback` multiple times effectively....

https://docs.reach.sh/frontend/#js_stdlib.setWalletFallback

**Comment by DanBurton** [comment]
When Jay says "it is now possible", he means, it will be possible starting with `@reach-sh/stdlib@0.1.11-rc.6`, which will be released next week.
